internal/alb/tg/targethealth.go

```diff
@@ -0,0 +1,339 @@
+func (c *targetHealthController) StopReconcilingPodConditionStatus(tgArn string) {
+	c.tgWatchesMux.Lock()
+	defer c.tgWatchesMux.Unlock()
+
+	tgWatch, ok := c.tgWatches[tgArn]
+	if ok {
+		tgWatch.cancel()
+		delete(c.tgWatches, tgArn)
+	}
+}
+
+// RemovePodConditions removes the condition (that we added before for the given ingress/backend) from the given pods
+func (c *targetHealthController) RemovePodConditions(ctx context.Context, t *Targets, targets []*elbv2.TargetDescription) error {
+	pods, err := c.endpointResolver.ReverseResolve(t.Ingress, t.Backend, targets)
+	if err != nil {
+		return err
+	}
+
+	conditionType := podConditionTypeForIngressBackend(t.Ingress, t.Backend)
+
+	for _, pod := range pods {
+		if pod != nil {
+			if i, cond := podConditionForReadinessGate(pod, conditionType); cond != nil {
+				pod.Status.Conditions = append(pod.Status.Conditions[:i], pod.Status.Conditions[i+1:]...)
+				err := c.client.Status().Update(ctx, pod)
+				if err != nil && !k8serrors.IsNotFound(err) {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// Background loop which keeps reconciling pod condition statuses for the given target groups until the given context is cancelled.
+func (c *targetHealthController) reconcilePodConditionsLoop(ctx context.Context, tgArn string, conditionType api.PodConditionType, tgWatch *targetGroupWatch) {
+	logger := albctx.GetLogger(ctx)
+	logger.Infof("Starting reconciliation of pod condition status for target group: %v", tgArn)
+
+	initial := true
+	for {
+		tgWatch.mux.Lock()
+		interval := c.ingressTargetHealthReconciliationInterval(initial, tgWatch.backend.ServiceName, tgWatch.ingress)
+		ingress := tgWatch.ingress
+		backend := tgWatch.backend
+		targetsToReconcile := append([]*elbv2.TargetDescription{}, tgWatch.targetsToReconcile...) // make copy
+		tgWatch.mux.Unlock()
+
+		initial = false
+
+		select {
+		case <-time.After(time.Duration(interval) * time.Second):
+			if err := c.reconcilePodConditions(ctx, tgArn, conditionType, ingress, backend, targetsToReconcile); err != nil {
+				logger.Errorf("Failed to reconcile pod condition status: %v", err)
+				albctx.GetEventf(ctx)(api.EventTypeWarning, "ERROR", "Error reconciling pod condition status via target group %s: %s", tgArn, err.Error())
+			}
+		case <-ctx.Done():
+			logger.Infof("Stopping reconciliation of pod condition status for target group: %v", tgArn)
+			return
+		}
+	}
+}
+
+// For each given pod, checks for the health status of the corresponding target in the target group and adds/updates a pod condition that can be used for pod readiness gates.
+func (c *targetHealthController) reconcilePodConditions(ctx context.Context, tgArn string, conditionType api.PodConditionType, ingress *extensions.Ingress, backend *extensions.IngressBackend, targetsToReconcile []*elbv2.TargetDescription) error {
+	in := &elbv2.DescribeTargetHealthInput{TargetGroupArn: aws.String(tgArn)}
+	resp, err := c.cloud.DescribeTargetHealthWithContext(ctx, in)
```

Comment:
DescribeTargetHealthInput supports passing targets as argument, which can safe calls & data transfer when there are lots of pod backends

Reply:
I see, I didn't notice that it had a parameter for targets. Will change it.